plugins/amongus/amongus.cpp

```diff
@@ -5,122 +5,132 @@
+									  "Context and identity are provided.";
 
-	identity = utf8ToUtf16(stream.str());
+	MumbleStringWrapper wrapper;
+	wrapper.data           = description;
+	wrapper.size           = sizeof(description) - 1;
+	wrapper.needsReleasing = false;
 
-	return true;
+	return wrapper;
 }
 
-static int tryLock(const std::multimap< std::wstring, unsigned long long int > &pids) {
-	const std::string name = "Among Us.exe";
-	const auto id          = Process::find(name, pids);
-	if (!id) {
-		return false;
-	}
-
-	game = std::make_unique< Game >(id, name);
-	if (!game->isOk()) {
-		game.reset();
-		return false;
-	}
+uint32_t mumble_getFeatures() {
+	return MUMBLE_FEATURE_POSITIONAL;
+}
 
-	const auto fields = game->clientFields();
-	if (!inGame(fields)) {
-		game.reset();
-		return false;
+uint8_t mumble_initPositionalData(const char *const *programNames, const uint64_t *programPIDs, size_t programCount) {
+	for (size_t i = 0; i < programCount; ++i) {
+		if (strcmp(programNames[i], "Among Us.exe") != 0) {
+			continue;
+		}
+
+		game = std::make_unique< Game >(programPIDs[i], programNames[i]);
+
+		const auto error = game->init();
+		if (error == Game::NONE) {
+			const auto fields = game->clientFields();
+			if (!game->isMultiplayer(fields)) {
+				game.reset();
+				return MUMBLE_PDEC_ERROR_TEMP;
+			}
+		} else {
+			game.reset();
+			return error == Game::TEMP ? MUMBLE_PDEC_ERROR_TEMP : MUMBLE_PDEC_ERROR_PERM;
+		}
+
+		return MUMBLE_PDEC_OK;
 	}
 
-	return true;
+	return MUMBLE_PDEC_ERROR_TEMP;
 }
 
-static const std::wstring longDesc() {
-	return std::wstring(L"Supports Among Us with context and identity support.");
+void mumble_shutdownPositionalData() {
+	game.reset();
 }
 
-static std::wstring description(L"Among Us");
-static std::wstring shortName(L"Among Us");
+bool mumble_fetchPositionalData(float *avatarPos, float *avatarDir, float *avatarAxis, float *cameraPos,
+								float *cameraDir, float *cameraAxis, const char **contextPtr,
+								const char **identityPtr) {
+	const auto fields = game->clientFields();
 
-static int tryLock1() {
-	return tryLock(std::multimap< std::wstring, unsigned long long int >());
-}
+	if (!game->isMultiplayer(fields)) {
+		return false;
+	}
 
-static void nullUnlock() {
-}
+	if (fields.gameState == GameState::Ended) {
+		return true;
+	}
 
-static MumblePlugin amongusPlug = { MUMBLE_PLUGIN_MAGIC, description, shortName, nullptr, nullptr, tryLock1,
-									nullUnlock,          longDesc,    fetch };
+	const auto playerControlFields = game->playerControlFields();
 
-static MumblePlugin2 amongusPlug2 = { MUMBLE_PLUGIN_MAGIC_2, MUMBLE_PLUGIN_VERSION, tryLock };
+	// Mumble | Game
+	// X      | X
+	// Y      | -
+	// Z      | Y
+	const auto position = game->playerPosition(playerControlFields);
+	avatarPos[0] = cameraPos[0] = position.x;
+	avatarPos[2] = cameraPos[2] = position.y;
 
-extern "C" MUMBLE_PLUGIN_EXPORT MumblePlugin *getMumblePlugin() {
-	return &amongusPlug;
-}
+	for (uint8_t i = 0; i < 3; ++i) {
```

Comment:
Using `uint8_t` instead of plain `unsigned int` might be a tiny performance hit for no apparent reason :eyes:

Reply:
Alternatively, I can use `uint_fast8_t` so that the architecture's behavior is taken into account.

However, we should make a clear decision and adhere to it in future code as well for consistency.

Reply:
I agree that a clear decision is desirable.

Given that we are also receiving PRs from third parties, I think that simply sticking to `int` or `unsigned int` makes life the easiest since that will probably be what most contributors will use.
Plus using fixed-width integer types always seems to suggest special meaning as in "this variable must for some reason have this exact width" which in cases like this does not apply :thinking:

Reply:
I would use `uint32_t` instead of `unsigned int`.

However, future architecture may be faster with 64 bit types, meaning that `uint_fast8_t` is probably the best choice.

Reply:
Here's the assembly output: https://godbolt.org/z/5TY98cxx9

Clang 12.0.0:
```asm
bit_8:                                  # @bit_8
        push    rbp
        mov     rbp, rsp
        mov     byte ptr [rbp - 17], 0
.LBB0_1:                                # =>This Inner Loop Header: Depth=1
        movzx   eax, byte ptr [rbp - 17]
        cmp     eax, 3
        jge     .LBB0_4
        movzx   eax, byte ptr [rbp - 17]
        xorps   xmm0, xmm0
        movss   dword ptr [rbp + 4*rax - 16], xmm0
        mov     al, byte ptr [rbp - 17]
        add     al, 1
        mov     byte ptr [rbp - 17], al
        jmp     .LBB0_1
.LBB0_4:
        mov     eax, dword ptr [rbp - 4]
        pop     rbp
        ret
bit_8_fast:                             # @bit_8_fast
        push    rbp
        mov     rbp, rsp
        mov     byte ptr [rbp - 17], 0
.LBB1_1:                                # =>This Inner Loop Header: Depth=1
        movzx   eax, byte ptr [rbp - 17]
        cmp     eax, 3
        jge     .LBB1_4
        movzx   eax, byte ptr [rbp - 17]
        xorps   xmm0, xmm0
        movss   dword ptr [rbp + 4*rax - 16], xmm0
        mov     al, byte ptr [rbp - 17]
        add     al, 1
        mov     byte ptr [rbp - 17], al
        jmp     .LBB1_1
.LBB1_4:
        mov     eax, dword ptr [rbp - 4]
        pop     rbp
        ret
bit_32:                                 # @bit_32
        push    rbp
        mov     rbp, rsp
        mov     dword ptr [rbp - 20], 0
.LBB2_1:                                # =>This Inner Loop Header: Depth=1
        cmp     dword ptr [rbp - 20], 3
        jae     .LBB2_4
        mov     eax, dword ptr [rbp - 20]
        xorps   xmm0, xmm0
        movss   dword ptr [rbp + 4*rax - 16], xmm0
        mov     eax, dword ptr [rbp - 20]
        add     eax, 1
        mov     dword ptr [rbp - 20], eax
        jmp     .LBB2_1
.LBB2_4:
        mov     eax, dword ptr [rbp - 4]
        pop     rbp
        ret
```

GCC 11.1
```asm
bit_8:
        push    rbp
        mov     rbp, rsp
        mov     BYTE PTR [rbp-1], 0
        jmp     .L2
.L3:
        movzx   eax, BYTE PTR [rbp-1]
        cdqe
        pxor    xmm0, xmm0
        movss   DWORD PTR [rbp-16+rax*4], xmm0
        add     BYTE PTR [rbp-1], 1
.L2:
        cmp     BYTE PTR [rbp-1], 2
        jbe     .L3
        nop
        pop     rbp
        ret
bit_8_fast:
        push    rbp
        mov     rbp, rsp
        mov     BYTE PTR [rbp-1], 0
        jmp     .L5
.L6:
        movzx   eax, BYTE PTR [rbp-1]
        cdqe
        pxor    xmm0, xmm0
        movss   DWORD PTR [rbp-16+rax*4], xmm0
        add     BYTE PTR [rbp-1], 1
.L5:
        cmp     BYTE PTR [rbp-1], 2
        jbe     .L6
        nop
        pop     rbp
        ret
bit_32:
        push    rbp
        mov     rbp, rsp
        mov     DWORD PTR [rbp-4], 0
        jmp     .L8
.L9:
        mov     eax, DWORD PTR [rbp-4]
        pxor    xmm0, xmm0
        movss   DWORD PTR [rbp-16+rax*4], xmm0
        add     DWORD PTR [rbp-4], 1
.L8:
        cmp     DWORD PTR [rbp-4], 2
        jbe     .L9
        nop
        pop     rbp
        ret
```

Reply:
> fixed-width integer types always seems to suggest special meaning as in "this variable must for some reason have this exact width" which in cases like this does not apply

I think this still is the most important argument as the performance implications are probably not measurable in like 99% of our cases.
But this argument also applies to `uint32_t` and `uint_fast8_t`. These types simply encode too much information that is completely irrelevant for situations like this.

> Here's the assembly output

That doesn't tell me anything as I don't know assembly :shrug:

Reply:
I'm no expert either, but I believe performance is exactly the same.